### ISSUE TITLE
[Bug][Spec Decode] probabilistic draft sampling corrupts output above max_num_seqs=2 at temperature > 0

### ISSUE THREAD

**Issue by Leonccaa** [1CatAI/1Cat-vLLM#656]
## Summary

With MTP speculative decoding and `draft_sample_method=probabilistic`, generated text is corrupted at the token level whenever `max_num_seqs > 2` and sampling temperature is above 0. Temperature 0 is unaffected, `greedy` is unaffected, and `max_num_seqs=2` is unaffected.

The failure is silent: no exception, no error counter, and the speculative acceptance rate stays in its normal range. Only the output is wrong.

Observed on SM70 (Tesla V100-PCIE-32GB x4, TP4, FP16 activations) with Qwen3.8-Flash-Next and `method=mtp`, `num_speculative_tokens=3`.

## Reproduction

The smallest reproduction changes exactly one field on an otherwise working deployment:

```
method=mtp, num_speculative_tokens=3, draft_sample_method=probabilistic
max_num_seqs: 2  ->  4
```

Everything else identical: same image, same checkpoint, same `gpu_memory_utilization`, FP16 (`auto`) KV cache, same prompt, same seed.

Prompt (Chinese, ~300 characters requested), `max_tokens=300`, `top_p` 1.0 at temperature 0.3 and 0.95 at 1.0:

| temperature | `max_num_seqs=2` | `max_num_seqs=4` |
| --- | --- | --- |
| 0.0 | clean | clean |
| 0.3 | clean | **corrupted** |
| 0.8 | clean | **corrupted** |
| 1.0 | clean | **corrupted** |

Representative corrupted output at `max_num_seqs=4`, temperature 0.3:

```
林婉的白天属于那套剪裁得体的深灰灰色西装。在会议室的冷白灯光下，她脊背挺得笔直，
指尖轻点桌面，语着PPT，眼神锐速平稳而精准，连发丝都端都像是不容置疑的秩序感。
```

Doubled characters (深灰灰色), spliced fragments (语着PPT, 眼神锐速平稳), and duplicated clause heads (都端都). At temperature 1.0 the output degrades much further. The same prompt and seed at `max_num_seqs=2` produces coherent prose.

## What it is not

Each of these was tested and ruled out by changing one variable at a time:

| hypothesis | test | result |
| --- | --- | --- |
| FP8 E4M3 KV cache | FP16 (`auto`) main and draft KV | still corrupted |
| E4M3 draft KV specifically | E4M3 main + FP16 draft | still corrupted |
| a recent code regression | the older production image, only `max_num_seqs` 2 -> 4 | still corrupted |
| checkpoint | a different checkpoint at `max_num_seqs=2` | clean |
| CUDA graph capture | `cudagraph_mode=PIECEWISE` at `max_num_seqs=4` | still corrupted |

Corruption tracks exactly two variables: `max_num_seqs > 2` **and** `draft_sample_method=probabilistic`. Eight configurations were tested and the two-factor rule explains all of them with no exceptions.

`greedy` at `max_num_seqs=4` with the same everything else is clean across temperatures 0.0/0.3/0.8/1.0, in 16 generations per concurrency across four prompts.

## Why it looks like an acceptance problem rather than a drafting one

Speculative-decoding counters for the same prompt, `max_num_seqs=4`:

| | temperature 0.0 | temperature 0.8 |
| --- | --- | --- |
| acceptance rate | 34.2% | 24.9% |
| mean accepted length | 2.03 | 1.75 |
| per-position accepted | 63 / 31 / 19 | 54 / 30 / 17 |

Acceptance is comparable between the clean and the corrupted case. Correct speculative sampling preserves the target distribution regardless of draft quality: a bad draft lowers the acceptance rate, it does not change what comes out. Wrong tokens are therefore being *accepted*, which points at the verification/correction step rather than at drafting.

Temperature 0 uses the greedy argmax match; temperature > 0 goes through the probabilistic rejection-sampling ratio test, which is the only path that reads the dense draft-logits buffer.

## Where we stopped

`Speculator.draft_logits` is allocated only for `draft_sample_method=probabilistic`, is shaped `(max_num_reqs, num_speculative_steps, vocab_size)`, is persistent across decode steps, and is initialised to a constant fill. Its first dimension is `max_num_seqs`, which is the variable that switches the bug on.

We checked that the write in `gumbel_block_argmax` and the read in the rejection-sampler Triton kernel both index it by `expanded_idx_mapping[...]`, i.e. the persistent request slot, so the indexing is symmetric at that level. Our working hypothesis is that some `(slot, draft step)` rows are read without having been written in that step, so the ratio test sees the fill value instead of a real distribution; that would explain wrong tokens being accepted at a normal-looking acceptance rate. **We did not instrument the buffer to confirm this**, so treat the localisation as a lead, not a diagnosis.

## Impact

`arg_utils.py` defaults MTP to `greedy`, so a default MTP deployment does not hit this. It is reachable by pinning `probabilistic` explicitly on MTP, which is how we found it.

`dflash` defaults to `probabilistic`, and the SM70 GLM-5.3 DFlash2 profiles in `config/vllm.py` require it. DFlash2 appears to be mostly insulated because its compact rejection path does not read the dense tensor, but the code comment on that buffer states that structured output and unsupported sampling contracts still fall back to the dense path. **We have not tested DFlash2**, so whether that fallback is exposed at `max_num_seqs > 2` is an open question for someone who owns that lane.

## Suggested mitigation until it is fixed

Either keep `max_num_seqs <= 2` when `draft_sample_method=probabilistic`, or reject the combination at startup so it fails loudly instead of producing plausible-looking wrong text.

## AI assistance

Investigation and this report were produced by Claude. The eight configurations above were each run on real hardware; the numbers are measured, not inferred.
